Implement percentage-closer soft shadows (PCSS).

pcwalton opened the pull request:
[*Percentage-closer soft shadows*] are a technique from 2004 that allow shadows to become blurrier farther from the objects that cast them. It works by introducing a *blocker search* step that runs before the normal shadow map sampling. The blocker search step detects the difference between the depth of the fragment being rasterized and the depth of the nearby samples in the depth buffer. Larger depth differences result in a larger penumbra and therefore a blurrier shadow.

To enable PCSS, fill in the `soft_shadow_size` value in `DirectionalLight`, `PointLight`, or `SpotLight`, as appropriate. This shadow size value represents the size of the light and should be tuned as appropriate for your scene.  Higher values result in a wider penumbra (i.e. blurrier shadows).

When using PCSS, temporal shadow maps (`ShadowFilteringMethod::Temporal`) are recommended. If you don't use `ShadowFilteringMethod::Temporal` and instead use `ShadowFilteringMethod::Gaussian`, Bevy will use the same technique as `Temporal`, but the result won't vary over time. This produces a rather noisy result. Doing better would likely require downsampling the shadow map, which would be complex and slower (and would require PR #13003 to land first).

In addition to PCSS, this commit makes the near Z plane for the shadow map configurable on a per-light basis. Previously, it had been hardcoded to 0.1 meters. This change was necessary to make the point light shadow map in the example look reasonable, as otherwise the shadows appeared far too aliased.

A new example, `pcss`, has been added. It demonstrates the percentage-closer soft shadow technique with directional lights, point lights, spot lights, non-temporal operation, and temporal operation. The assets are my original work.

Both temporal and non-temporal shadows are rather noisy in the example, and, as mentioned before, this is unavoidable without downsampling the depth buffer, which we can't do yet. Note also that the shadows don't look particularly great for point lights; the example simply isn't an ideal scene for them. Nevertheless, I felt that the benefits of the ability to do a side-by-side comparison of directional and point lights outweighed the unsightliness of the point light shadows in that example, so I kept the point light feature in.

Fixes #3631.

[*Percentage-closer soft shadows*]:
https://developer.download.nvidia.com/shaderlibrary/docs/shadow_PCSS.pdf

## Changelog

### Added

* Percentage-closer soft shadows (PCSS) are now supported, allowing shadows to become blurrier as they stretch away from objects. To use them, set the `soft_shadow_size` field in `DirectionalLight`, `PointLight`, or `SpotLight`, as applicable.

* The near Z value for shadow maps is now customizable via the `shadow_map_near_z` field in `DirectionalLight`, `PointLight`, and `SpotLight`.

## Screenshots

PCSS off:
![Screenshot 2024-05-24 120012](https://github.com/bevyengine/bevy/assets/157897/0d35fe98-245b-44fb-8a43-8d0272a73b86)

PCSS on:
![Screenshot 2024-05-24 115959](https://github.com/bevyengine/bevy/assets/157897/83397ef8-1317-49dd-bfb3-f8286d7610cd)

torsteingrindvik commented on crates/bevy_pbr/src/light/spot_light.rs:
```diff
@@ -7,23 +7,78 @@ use super::*;
+
+    /// Range in meters that this light illuminates.
+    ///
+    /// Note that this value affects resolution the shadow maps; generally, the
```
```suggestion
    /// Note that this value affects resolution of the shadow maps; generally, the
```

I think the intention was?

torsteingrindvik commented on crates/bevy_pbr/src/light/spot_light.rs:
```diff
@@ -7,23 +7,78 @@ use super::*;
     /// away.
     pub shadow_normal_bias: f32,
+
+    /// The distance from the light to near Z plane in the shadow map.
```
```suggestion
    /// The distance from the light to the near Z plane in the shadow map.
```

torsteingrindvik commented:
I don't know the domain well- why do the shadows here: 
![image](https://github.com/bevyengine/bevy/assets/52322338/57210e86-2bc5-46e5-ae3d-e080bbf0d6d6)

Have curve-like imperfections?

I noticed the article you linked has this image which has a different type of noise:

![image](https://github.com/bevyengine/bevy/assets/52322338/3a336745-4ed9-45eb-8708-443fb2ddeac2)

Is it related to the geometry which casts the shadow or something else?


Also I was wondering how the shadows perform when the light source is moving.
I changed the example a bit to have the light undulate.

[pcss-example-undulate.patch.txt](https://github.com/bevyengine/bevy/files/15443178/pcss-example-undulate.patch.txt)

Feel free to `git apply pcss-example-undulate.patch.txt` if you want to try the same or add it to the PR.


https://github.com/bevyengine/bevy/assets/52322338/2a6530dd-c0f8-4fde-b6c7-fa901716b128


I noticed an effect that spot lights have this effect where it looks like the shadow density changes a lot, perhaps related to shadow map resolution?

pcwalton commented:
The patterns are because of the [interleaved gradient noise](https://blog.demofox.org/2022/01/01/interleaved-gradient-noise-a-different-kind-of-low-discrepancy-sequence/). This is generally considered more desirable than the white noise that it looks like your other example has.

There isn't a whole lot we can do about the noise in this patch without generating mip chains for the shadow map, which would require https://github.com/bevyengine/bevy/pull/13003 (and even if that were in I wouldn't want to do it in the first patch because this patch is too big as it is). When I asked in Discord I was told that PCSS in games tends to be noisy; that's just the way it is.

pcwalton commented:
I considered allowing the user to move the light but I realized that this will cause problems with the Z near value, which is carefully tuned so that the palm tree makes maximum use of the shadow map. So I'm less inclined to make the light movable, though I could be convinced otherwise.

pcwalton commented:
I suspect what's going wrong with spot lights is that the blocker search starts missing blockers. I'm not sure if this is something that can reasonably be fixed; it's always possible that the blocker search can miss things. The solution is for artists to tweak the `soft_shadow_size` manually.

torsteingrindvik reviewed (approved):
I found out that Godot has PCSS and they have an example showcasing it: https://github.com/godotengine/godot-demo-projects/tree/master/3d/lights_and_shadows

I put the GLTF files in this PR in that project and dropped the palm tree into Godot and tried experimenting with the PCSS lights in that example.
I couldn't try the spot light (likely due to https://github.com/godotengine/godot/issues/80137) but I saw similar results when using their omni light (point light) and their sun (directional light).

I learned that it's very finicky to get decent looking shadows, and I had to resort to a mix of tuning their shadow blur and their shadow opacity (not sure if we have that in Bevy yet) and only specific combos of those along with light strength and positioning yielded OK results.

I'm convinced that PCSS is hard to get right (as a user) and that this PR adds value and if we can get better noise patterns further down the line that's really great.

EDIT: When we have a Bevy editor where we can interactively tune parameters like I could in Godot it will be a lot easier to spot problems so I'm not worried about any minor issues right now.

Approving on the basis of the above, as well as:

- Code quality to me looks great
- Adds great docs
- Great example

I haven't verified any math or looked in depth at the shaders- hoping rendering maintainers will do so.

pcwalton commented:
Yeah. Unity doesn't implement PCSS at all and in this PR I pretty much learned why. PCSS is just really hard to tune.

theprotonfactor commented:
@pcwalton Unity does, in fact, implement PCSS as part of their High Definition Render Pipeline package. See the "Shadows" section on [this](https://github.com/Unity-Technologies/Graphics/blob/master/Packages/com.unity.render-pipelines.high-definition/Documentation~/whats-new-16.md)   changelog page, for example. If you ever want to know what Unity does and doesn't support then that repository is a fantastic place to search.

therealbnut commented on crates/bevy_pbr/src/cluster/mod.rs:
```diff
@@ -151,6 +151,10 @@ pub struct GpuClusterableObject {
     pub(crate) spot_light_tan_angle: f32,
+    pub(crate) soft_shadow_size: f32,
+    pub(crate) shadow_map_near_z: f32,
+    pub(crate) pad_a: f32,
+    pub(crate) pad_b: f32,
```
Nit: Is this intended to make the structure's size a multiple of 4 x f32? If so a comment and maybe a test might be good, assuming it's important and relied upon elsewhere. Maybe something like this would also make the intention clearer and more maintainable:
```suggestion
    pub(crate) _reserved: [f32; 2],
```

The test could be as simple as this:
```rust
#[cfg(test)]
const_assert_eq!(size_of::<GpuClusterableObject>(), size_of::<Vec4>());
```

IceSentry replied:
A test isn't that necessary for that. Things break pretty quickly if the padding is wrong anyway. We have padding in a bunch of places for rendering code because of some webgl limitations.

therealbnut replied:
Cool, no problem if it’s convention - although should it be commented all the same?

IceSentry replied:
For me, just the fact that it's called `pad_x` is enough documentation.

therealbnut replied:
Righto, again no problem if it's an established convention in the codebase.

I was initially unsure whether this was perhaps something like padding from the edge of the shadow when sampling to avoid boundary conditions. It required me to check it wasn't used to understand its purpose and intention, but if I was adding a new field I may not have checked so thoroughly and just added new fields rather than removing these.

Again, it's a nit, and not a huge issue. I just saw it as something that could be made clearer. My intention if it's clearer is that more people can get across these changes, review, and contribute. Being knowledgable about graphics doesn't mean you're knowledgable about conventions in this codebase, or rust, etc. My understanding of Rust convention, for example, is that fields like this would have an underscore prefix `_pad_a` to indicate they're intentionally unused.

therealbnut reviewed (approved):
Code looks good, just a small nit about clarifying some unused uncommented fields.

pcwalton commented:
Please don't merge this yet as spot shadows are broken somehow.

therealbnut commented on crates/bevy_pbr/src/render/mesh_view_bindings.rs:
```diff
@@ -213,11 +213,13 @@ fn layout_entries(
+            // Point Shadow Texture Array Linear Sampler
+            (4, sampler(SamplerBindingType::Filtering)),
             // Directional Shadow Texture Array
             (
-                4,
+                5,
```
Not a change for this PR, but after this is merged would a PR be welcome to replace these values with constants? and defined in the shader with defines like this:
https://github.com/bevyengine/bevy/blob/4a312a12e79ccf32b54ba2d3e99053685028c055/crates/bevy_core_pipeline/src/tonemapping/lut_bindings.wgsl#L3-L4

It might reduce merge conflict churn when there's lots of unrelated changes in a PR.

pcwalton replied:
As I recall, non-boolean constants are kind of annoying in our PBR shaders at the moment, but sure, it could be done in the future.

pcwalton commented:
This is ready for review again. I fixed the regression. I also moved the radio button code into a reusable helper module. That way, I don't have to keep recreating it for every new example.

pcwalton commented on examples/3d/pcss.rs:
```diff
@@ -0,0 +1,641 @@
+};
+
+/// The path to the UI font.
+static FONT_PATH: &str = "fonts/FiraMono-Medium.ttf";
```
It's removed.

IceSentry reviewed (approved):
I tested a bunch of examples and didn't find any regression.

LGTM

superdump commented on crates/bevy_pbr/src/light/directional_light.rs:
```diff
@@ -71,6 +110,7 @@ impl Default for DirectionalLight {
             color: Color::WHITE,
             illuminance: light_consts::lux::AMBIENT_DAYLIGHT,
             shadows_enabled: false,
+            soft_shadow_size: None,
```
Could we add a constant to make it easy to set the Sun's diameter?

pcwalton replied:
I believe that using the sun's diameter here would not be particularly useful unless the sun were its actual physical distance away from the target, which would be impractical.

superdump commented on crates/bevy_pbr/src/light/mod.rs:
```diff
@@ -586,6 +579,16 @@ pub fn update_point_light_frusta(
+            1.0,
+            point_light.shadow_map_near_z,
+        );
+        let view_rotations = CUBE_MAP_FACES
+            .iter()
+            .map(|CubeMapFace { target, up }| Transform::IDENTITY.looking_at(*target, *up))
+            .collect::<Vec<_>>();
```
Should this be moved in here? They don't change dependent on the point light?

pcwalton replied:
Reverted the change.

superdump commented on crates/bevy_pbr/src/light/point_light.rs:
```diff
@@ -22,29 +22,60 @@ use super::*;
+    ///
+    /// Note that soft shadows are significantly more expensive to render than
+    /// hard shadows.
+    pub soft_shadow_size: Option<f32>,
```
Is there a reason to use a separate field for the size of the light source here when there is already a `radius` field? I suppose if gives more not-physically-correct artistic control over the shadow softness independent of the specular highlights.

pcwalton replied:
I removed the separate field in favor of just using `radius`.

superdump commented on crates/bevy_pbr/src/render/light.rs:
```diff
@@ -875,6 +891,16 @@ pub fn prepare_lights(
+                1.0,
+                light.shadow_map_near_z,
+            );
+            let cube_face_rotations = CUBE_MAP_FACES
+                .iter()
+                .map(|CubeMapFace { target, up }| Transform::IDENTITY.looking_at(*target, *up))
+                .collect::<Vec<_>>();
```
No dependence on anything in this inner loop so could remain where it was?

pcwalton replied:
Reverted the change.

superdump commented on examples/3d/pcss.rs:
```diff
@@ -0,0 +1,424 @@
+#[path = "../helpers/widgets.rs"]
+mod widgets;
+
+/// The size of the soft shadow penumbras when PCSS is enabled.
```
The penumbra scales according to ratios of distances between the receiving surface, the occluder, the light source, and the light source's radius. As such, this seems incorrect and misleading?

pcwalton replied:
I renamed this to `LIGHT_RADIUS` and reworded the comment.

Olle-Lukowski reviewed (approved):
Looks great! I am wondering though, are there any other ways we can make it play nicely with a point light? Or is that a whole other technique?

alice-i-cecile commented:
@pcwalton I'm inclined to merge this basically as is. When you have a moment, could you clean up merge conflicts? I don't feel confident doing so myself here.

@superdump's comments are solid as well, but I don't feel that they're ultimately blocking (feel free to correct me).

pcwalton commented:
@Olle-Lukowski It does work with point lights. This just isn't the prettiest scene for them.

pcwalton commented:
Rebased and comments addressed. I think this is OK to go unless it needs a rereview.

alice-i-cecile commented:
DirectX12 failures: https://github.com/bevyengine/bevy/actions/runs/10909293058/job/30277186297

pcwalton commented:
@alice-i-cecile I think I fixed DX12.

pcwalton commented:
This failure is because `directional_shadow_textures` can be sampled with both linear and PCF with PCSS and WebGL 2 doesn't support that. Grr. Trying to see if there's a way I can just ifdef off all of PCSS in WebGL 2.

pcwalton commented:
Should be fixed again.

alice-i-cecile commented:
I love WebGL2 🥲 Thanks for fixing this up <3

alice-i-cecile commented:
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1688 if you'd like to help out.

magcius commented on crates/bevy_pbr/src/render/shadow_sampling.wgsl:
```diff
@@ -169,17 +264,100 @@ fn sample_shadow_map(light_local: vec2<f32>, depth: f32, array_index: i32, texel
+    let z_blocker = search_for_blockers_in_shadow_map(
+        light_local, depth, array_index, texel_size, light_size);
+
+    // Don't let the blur size go below 0.5, or shadows will look unacceptably aliased.
+    let blur_size = max((z_blocker - depth) * light_size / depth, 0.5);
```
This formula isn't quite right. The original formula is based off of similar triangles/ratios: if you have an area light of size S, and the distance to the blocker is D, that's one triangle. Then you construct the other triangle by taking the distance from the blocker to the receiver, and the penumbra on the ground is the ratio.

Thus we end up with `light_size * (z_receiver - z_blocker) / z_blocker`; however, this only works if the cascade is a cube (so that we can take distances along Z and use them to calculate distances along X/Y), and it assumes that the depths are distances from the light, rather than inverted Z-buffer values like they are here.

We could fix both of these things, but this formulation has a bigger issue for sunlights, which is that close cascades and pancaking really screws this up; any blocker which is beyond the cascade will be at the light's near plane, which means we have a massive penumbra. This is a huge distortion, and unfortunately different cascades will have different distortions, so they're tricky to blend together. Effectively, the classic PCSS formula tries to relate the distance between the light and the blocker to the distance between the blocker and the receiver, but the sunlight is infinitely far away.

The resulting formula that you ended up here with is nonphysical, but more stable, and honestly fine. Effectively, you're scaling the blur amount based on the distance between the blocker and the receiver, where bigger distances means larger penumbras, using `light_size` as the maximum penumbra size.

The only recommendation here I'd make is to drop the `/ depth`; this is nonphysical and I suspect only really looks OK because the depth differentials on receivers is going to be relatively small.

Also, spotlights don't have either issue (there's only one cascade, and no pancaking of casters, all casters are going to be in front of the light), so the spotlight PCSS code could be accurate if wanted, but it also might make sense to match spotlights with sunlights just so the two behaved similarly.
